File: tools/check-yaml-interop.js
// Verify Section 19.4's emitted YAML against an independent YAML 1.2 reader.
//
// Section 19.4 quotes a scalar that either published YAML schema would type. The Python lane
// covers YAML 1.1 through PyYAML; this one covers YAML 1.2 through js-yaml, and the two are
// not interchangeable. A 1.1 reader types `yes`, `12:30`, `2001-12-14` and `1_000` and leaves
// `0o17` a string; a 1.2 reader does the reverse. Either lane alone would pass a writer that
// picked one revision and left the other's readers wrong, which is the whole reason Section
// 19.4 takes the union rather than a side.
//
// The oracle is the corpus, not the implementation. For a case whose input is JSON and whose
// scheme only names an output format, the emitted YAML must carry exactly the document the
// JSON carried: JSON.parse fixes the types with no YAML resolution involved, so comparing the
// two establishes that a third-party reader gives the values back rather than merely
// accepting the file.
//
// Nothing here imports the implementation, and nothing here re-implements Section 19.4's
// spelling rules.
//
// Usage:  node tools/check-yaml-interop.js [repo-root]

'use strict';

const fs = require('fs');
const path = require('path');

let yaml;
try {
  yaml = require('js-yaml');
} catch (error) {
  process.stderr.write('check-yaml-interop: js-yaml is not installed\n');
  process.exit(2);
}

// Cases whose emitted bytes carry the spellings the two YAML revisions disagree about. The
// lane is nearly vacuous without them, so their absence from the checked set is a failure
// rather than a skip.
const REQUIRED_CASES = [
  'yaml-quotes-every-portably-typed-spelling',
  'yaml-scalar-style-selection',
];

// The shared selector an output-only scheme names, or null when it names more. A scheme that
// filters or transforms makes the emitted document a function of rules this lane does not
// model. One shared selector is modelled: the emitted document is then the subtree the JSON
// carries at that path.
function schemePrefix(text) {
  const prefixes = new Set();
  for (const raw of text.split('\n')) {
    const line = raw.trim();
    if (line === '' || line.startsWith('#')) {
      continue;
    }
    const selector = line.split('=')[0].trim();
    const cut = selector.lastIndexOf('.');
    const head = cut === -1 ? '' : selector.slice(0, cut);
    const name = cut === -1 ? selector : selector.slice(cut + 1);
    if (name.toLowerCase() !== 'output' && name.toLowerCase() !== 'filename') {
      return null;
    }
    prefixes.add(head);
  }
  if (prefixes.size !== 1) {
    return null;
  }
  const only = [...prefixes][0];
  return only === '' ? [] : only.split('.');
}

function listFiles(dir, extension) {
  if (!fs.existsSync(dir)) {
    return [];
  }
  return fs.readdirSync(dir).filter((name) => name.endsWith(extension)).sort()
    .map((name) => path.join(dir, name));
}

// Structural equality over the shapes JSON and YAML share. Key order is not compared: the
// corpus pins order in bytes, and this lane is about values surviving a reader.
function same(left, right) {
  if (left === null || right === null) {
    return left === right;
  }
  if (Array.isArray(left) || Array.isArray(right)) {
    return Array.isArray(left) && Array.isArray(right)
      && left.length === right.length
      && left.every((item, index) => same(item, right[index]));
  }
  if (typeof left === 'object' || typeof right === 'object') {
    if (typeof left !== 'object' || typeof right !== 'object') {
      return false;
    }
    const leftKeys = Object.keys(left).sort();
    const rightKeys = Object.keys(right).sort();
    return leftKeys.length === rightKeys.length
      && leftKeys.every((key, index) => key === rightKeys[index])
      && leftKeys.every((key) => same(left[key], right[key]));
  }
  return Object.is(left, right);
}

// A key a 1.2 reader did not give back as a string. js-yaml stringifies object keys, so a
// resolved Boolean or number arrives as its canonical text rather than the written spelling;
// comparing the two is what exposes the collapse.
function nonStringKeys(node, written, found) {
  if (node === null || typeof node !== 'object') {
    return;
  }
  if (Array.isArray(node)) {
    node.forEach((item) => nonStringKeys(item, written, found));
    return;
  }
  for (const key of Object.keys(node)) {
    if (!written.has(key)) {
      found.push(key);
    }
    nonStringKeys(node[key], written, found);
  }
}

function main(argv) {
  const root = path.resolve(argv[2] || '.');
  const conformance = path.join(root, 'conformance');
  const failures = [];
  const checked = new Set();
  let loaded = 0;

  for (const caseName of fs.readdirSync(conformance).sort()) {
    const caseDir = path.join(conformance, caseName);
    if (!fs.statSync(caseDir).isDirectory()) {
      continue;
    }

    for (const emitted of listFiles(path.join(caseDir, 'expected'), '.yaml')) {
      loaded += 1;
      try {
        yaml.load(fs.readFileSync(emitted, 'utf8'));
      } catch (error) {
        failures.push(`${path.relative(root, emitted)}: a YAML 1.2 reader refuses the document -- ${error.message.split('\n')[0]}`);
      }
    }

    const inputs = listFiles(path.join(caseDir, 'inputs'), '.json');
    const emittedYaml = listFiles(path.join(caseDir, 'expected'), '.yaml');
    const schemes = listFiles(path.join(caseDir, 'schemes'), '.txt');
    if (inputs.length !== 1 || emittedYaml.length !== 1 || schemes.length !== 1) {
      continue;
    }
    const prefix = schemePrefix(fs.readFileSync(schemes[0], 'utf8'));
    if (prefix === null) {
      continue;
    }

    let expected = JSON.parse(fs.readFileSync(inputs[0], 'utf8'));
    let reachable = true;
    for (const component of prefix) {
      if (expected === null || typeof expected !== 'object' || !(component in expected)) {
        reachable = false;
        break;
      }
      expected = expected[component];
    }
    if (!reachable) {
      continue;
    }

    let actual;
    try {
      actual = yaml.load(fs.readFileSync(emittedYaml[0], 'utf8'));
    } catch (error) {
      checked.add(caseName);
      continue;
    }

    if (!same(actual, expected)) {
      failures.push(`${caseName}: '${path.basename(emittedYaml[0])}' does not read back as the JSON it was built from\n`
        + `      json: ${JSON.stringify(expected)}\n      yaml: ${JSON.stringify(actual)}`);
    }

    const written = new Set();
    (function collect(node) {
      if (node === null || typeof node !== 'object') {
        return;
      }
      if (Array.isArray(node)) {
        node.forEach(collect);
        return;
      }
      Object.keys(node).forEach((key) => { written.add(key); collect(node[key]); });
    }(expected));

    const strayKeys = [];
    nonStringKeys(actual, written, strayKeys);
    for (const key of strayKeys) {
      failures.push(`${caseName}: key '${key}' is not a key the JSON wrote, so a reader resolved it away from its spelling`);
    }

    checked.add(caseName);
  }

  const missing = REQUIRED_CASES.filter((name) => !checked.has(name));
  if (missing.length > 0) {
    failures.push(`the cases this lane exists for were not compared: ${missing.sort().join(', ')}`);
  }

  if (failures.length > 0) {
    process.stderr.write('check-yaml-interop: FAILED\n');
    for (const failure of failures) {
      process.stderr.write(`  - ${failure}\n`);
    }
    return 1;
  }

  process.stdout.write(`check-yaml-interop: ${loaded} emitted YAML files load under js-yaml; `
    + `${checked.size} compared to the JSON they were built from\n`);
  return 0;
}

process.exit(main(process.argv));

File: tools/check-yaml-interop.py
"""Verify Section 19.4's emitted YAML against an independent YAML 1.1 reader.

Section 19.4 quotes a scalar that either published YAML schema would type, so that a reader
returns the string that was written whichever revision its library implements. This is that
test for PyYAML, whose resolver is YAML 1.1 -- the stricter of the two, and the one that
types `yes`, `12:30`, `2001-12-14`, `1_000` and the two key tags.

The oracle is the corpus, not the implementation. For a case whose input is JSON and whose
scheme only names an output format, the emitted YAML must carry exactly the document the
JSON carried: `json.load` fixes the types with no YAML resolution involved, so comparing the
two establishes that a third-party reader gives the values back rather than merely accepting
the file. That is the distinction the M5 review turned on -- the tool's own reader shares the
tool's own resolutions, so a same-format round trip through Section 3.3 succeeds while every
other reader disagrees, and the failure is invisible from inside.

Nothing here imports the implementation, and nothing here re-implements Section 19.4's
spelling rules. A defect in the writer therefore cannot define its own oracle.

The broad lane adds a weaker check over every emitted YAML file in the corpus: the document
must load at all, and every mapping key must come back as a string. Both catch losses that
change no byte the corpus compares -- a document a reader abandons on an unresolvable tag,
and two keys distinct as written that resolve to one.

Usage:  python tools/check-yaml-interop.py [repo-root]
"""

import json
import pathlib
import sys

try:
    import yaml
except ImportError:  # pragma: no cover - the lane installs it
    print("check-yaml-interop: PyYAML is not installed", file=sys.stderr)
    raise SystemExit(2)

# Cases whose emitted bytes carry the spellings the two YAML revisions disagree about. The
# lane is nearly vacuous without them -- every other emitted YAML file in the corpus is
# lowercase ASCII words -- so their absence from the checked set is a failure rather than a
# skip. A check that stops exercising the thing it was built for should say so.
REQUIRED_CASES = (
    "yaml-quotes-every-portably-typed-spelling",
    "yaml-scalar-style-selection",
)


def scheme_prefix(scheme_text):
    """The single selector an output-only scheme names, or None when it names more.

    A scheme that filters, transforms values, or spreads its directives over more than one
    selector makes the emitted document a function of rules this lane does not model, so the
    strong comparison does not apply. One shared selector is modelled: the emitted document
    is then the subtree the JSON carries at that path.

    :param scheme_text: The scheme file's text.
    :returns: The shared selector as a list of components, or None when ineligible.
    """
    prefixes = set()
    for raw in scheme_text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        selector = line.split("=", 1)[0].strip()
        head, _, name = selector.rpartition(".")
        if name.lower() not in ("output", "filename"):
            return None
        prefixes.add(head)
    if len(prefixes) != 1:
        return None
    only = prefixes.pop()
    return only.split(".") if only else []


def strong_lane(root, failures):
    """Compare each eligible case's emitted YAML to the JSON it was built from.

    :param root: The repository root.
    :param failures: A list collecting failure descriptions.
    :returns: The set of case names actually compared.
    """
    checked = set()
    for case in sorted((root / "conformance").iterdir()):
        if not case.is_dir():
            continue
        inputs = sorted((case / "inputs").glob("*.json")) if (case / "inputs").is_dir() else []
        emitted = sorted((case / "expected").glob("*.yaml")) if (case / "expected").is_dir() else []
        schemes = sorted((case / "schemes").glob("*.txt")) if (case / "schemes").is_dir() else []
        if len(inputs) != 1 or len(emitted) != 1 or len(schemes) != 1:
            continue
        prefix = scheme_prefix(schemes[0].read_text(encoding="utf-8"))
        if prefix is None:
            continue

        expected = json.loads(inputs[0].read_text(encoding="utf-8"))
        for component in prefix:
            if not isinstance(expected, dict) or component not in expected:
                expected = None
                break
            expected = expected[component]
        if expected is None:
            continue

        try:
            actual = yaml.safe_load(emitted[0].read_text(encoding="utf-8"))
        except Exception as error:  # noqa: BLE001 - any refusal is the finding
            failures.append(
                "%s: a YAML 1.1 reader abandons '%s' -- %s: %s"
                % (case.name, emitted[0].name, type(error).__name__, str(error).splitlines()[0])
            )
            checked.add(case.name)
            continue

        if actual != expected:
            failures.append(
                "%s: '%s' does not read back as the JSON it was built from\n"
                "      json: %r\n      yaml: %r" % (case.name, emitted[0].name, expected, actual)
            )
        checked.add(case.name)
    return checked


def broad_lane(root, failures):
    """Load every emitted YAML file and check that no key resolves away from a string.

    :param root: The repository root.
    :param failures: A list collecting failure descriptions.
    :returns: The number of files loaded.
    """
    count = 0
    for path in sorted((root / "conformance").glob("*/expected/*.yaml")):
        count += 1
        try:
            document = yaml.safe_load(path.read_text(encoding="utf-8"))
        except Exception as error:  # noqa: BLE001 - any refusal is the finding
            failures.append(
                "%s: a YAML 1.1 reader abandons the document -- %s: %s"
                % (path.relative_to(root), type(error).__name__, str(error).splitlines()[0])
            )
            continue
        for where, key in non_string_keys(document, path.relative_to(root)):
            failures.append("%s: key %r came back as %s, not a string" % (where, key, type(key).__name__))
    return count


def non_string_keys(node, where):
    """Yield every mapping key that did not come back as a string.

    A component name is text, so a key that resolves to a Boolean, a number, or a date is a
    key two distinct spellings can collide on -- the loss Section 19.3's `FLAT001` prevents
    from inside the tool and quoting prevents from outside it.

    :param node: The loaded document or a node within it.
    :param where: The file the node came from, for the message.
    :returns: Pairs of location and offending key.
    """
    if isinstance(node, dict):
        for key, value in node.items():
            if not isinstance(key, str):
                yield where, key
            yield from non_string_keys(value, where)
    elif isinstance(node, list):
        for item in node:
            yield from non_string_keys(item, where)


def main(argv):
    """Run both lanes and report.

    :param argv: Command line arguments; an optional repository root.
    :returns: The process exit code.
    """
    root = pathlib.Path(argv[1] if len(argv) > 1 else ".").resolve()
    failures = []

    checked = strong_lane(root, failures)
    loaded = broad_lane(root, failures)

    missing = [case for case in REQUIRED_CASES if case not in checked]
    if missing:
        failures.append(
            "the cases this lane exists for were not compared: " + ", ".join(sorted(missing))
        )

    if failures:
        print("check-yaml-interop: FAILED", file=sys.stderr)
        for failure in failures:
            print("  - " + failure, file=sys.stderr)
        return 1

    print(
        "check-yaml-interop: %d emitted YAML files load under PyYAML; "
        "%d compared to the JSON they were built from" % (loaded, len(checked))
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
